fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/FunctionBuilder.java

```diff
@@ -35,12 +36,43 @@
 public class FunctionBuilder {
     public final int arity;
 
+    public final boolean isVariableLength;
+
     // Concrete BoundFunction's constructor
     private final Constructor<BoundFunction> builderMethod;
 
     public FunctionBuilder(Constructor<BoundFunction> builderMethod) {
         this.builderMethod = Objects.requireNonNull(builderMethod, "builderMethod can not be null");
         this.arity = builderMethod.getParameterCount();
+        this.isVariableLength = arity > 0 && builderMethod.getParameterTypes()[arity - 1].isArray();
+    }
+
+    /** check whether arguments can apply to the constructor */
+    public boolean canApply(List<? extends Object> arguments) {
+        if (isVariableLength && arity > arguments.size() + 1) {
+            return false;
+        }
+        if (!isVariableLength && arguments.size() != arity) {
+            return false;
+        }
+        for (int i = 0; i < arguments.size(); i++) {
+            Class constructorArgumentType = getConstructorArgumentType(i);
+            if (!constructorArgumentType.isInstance(arguments.get(i))) {
```

Comment:
We should check if the class is precisely the same or only an instance of (derived class ok) ?

Reply:
This code say it can apply if the argument type is the same as signature type or is the child class of the signature type.
For example, `Substring(StringLiteral, IntegerLiteral, IntegerLiteral)` can apply to `Substring(Expression, Expression, Expression)`